packages/cdktf-cli/lib/server/WatchClient.ts

```diff
@@ -0,0 +1,353 @@
+        const { planFile } = await terraform.plan(false); // false = deploy
+        await terraform.deploy(planFile, this.handleTerraformOutput.bind(this));
+      } else {
+        // skip the plan to save time ⏱
+        const NO_PLAN_FILE = "";
+        await terraform.deploy(
+          NO_PLAN_FILE,
+          this.handleTerraformOutput.bind(this)
+        );
+      }
+      // deployment was successful -> update hash
+      this.lastTargetStackHash = newTargetStackHash;
+    } catch (e) {
+      // When new providers where added or the stack name might've
+      // been changed (when using just a single stack) we need to
+      // re-init Terraform. We'll do this proactively on any error.
+      this.needsInit = true;
+      throw new RecoverableError(
+        `${e.message}${e.stderr ? `\n${e.stderr}` : ""}`,
+        "DEPLOYING"
+      );
+    }
+  }
+
+  private handleTerraformOutput(chunk: Buffer) {
+    const resources = parseOutput(chunk.toString());
+    if (resources.length > 0) {
+      const newResources = new Map(
+        this.state.resources as WatchState["resources"]
+      );
+      resources.forEach((r) =>
+        newResources.set(r.id, { ...r, changedAt: Date.now() })
+      );
+      this.updateState({ resources: newResources as WatchState["resources"] });
+    }
+  }
+
+  private async getTargetStack(): Promise<Stack> {
+    const { stacks } = this.state;
+    if (stacks.length === 0) {
+      throw new RecoverableError(
+        "Cannot determine target stack when there are no stacks",
+        this.state.status
+      );
+    }
+    let stack: Stack | undefined = stacks[0];
+    if (this.targetStack) {
+      stack = stacks.find((s) => s.name === this.targetStack);
+      if (!stack) {
+        throw new RecoverableError(
+          `Could not find stack ${this.targetStack}. Found ${stacks
+            .map((s) => s.name)
+            .join(", ")}`,
+          this.state.status
+        );
+      }
+    } else if (stacks.length > 1) {
+      throw new RecoverableError(
+        "Found more than one stack, please specify which stack to watch using --stack",
+        this.state.status
+      );
+    }
+    return stack;
+  }
+
+  // todo: optimization: cache instance as long as backend does not change
+  private async getTerraform(): Promise<Terraform> {
+    const stack = await this.getTargetStack();
+    if (stack.json.terraform?.backend?.remote) {
+      const terraformCloud = new TerraformCloud(
+        stack,
+        stack.json.terraform?.backend?.remote
+      );
+      if (await terraformCloud.isRemoteWorkspace()) {
+        throw new RecoverableError(
+          "Using Terraform Cloud in Remote Execution Mode is not yet supported for cdktf watch",
+          this.state.status
+        );
+        // return terraformCloud;
+      }
+    }
+    return new TerraformCli(stack);
+  }
+
+  public async start() {
+    this.running = true;
+
+    const gitignored = await readGitignore(process.cwd());
```

Comment:
Nitpick: There is no cheap way to listen to gitignores as well and reevaluate this, right? I'm always annoyed when I have to restart sth due to such a config change

Reply:
Comes with complexity that I just wanted to avoid for the first version. Another thing that we currently also don't handle are nested `.gitignore` files.